libglusterfs/src/common-utils.c

```diff
@@ -642,6 +642,123 @@ gf_dnscache_entry_deinit(struct dnscache_entry *entry)
+        gf_msg_plain(GF_LOG_WARNING, "dnscache is NULL");
+        return;
+    }
+
+    dns_addr_pairs = cache->cache_dict->members_list;
+    while (dns_addr_pairs) {
+        dnsentry = (struct dnscache_addr_entry *)dns_addr_pairs->value->data;
+        freeaddrinfo(dnsentry->addr);
+        dns_addr_pairs = dns_addr_pairs->next;
+    }
+    gf_dnscache_deinit(cache);
+}
+
+/**
+ * gf_dnscache_entry_init -- Initialize a dnscache entry
+ *
+ * @return: SUCCESS: Pointer to an allocated dnscache entry struct
+ *          FAILURE: NULL
+ */
+
+struct dnscache_addr_entry *
+gf_dnscache_addr_entry_init()
+{
+    struct dnscache_addr_entry *entry = GF_CALLOC(
+        1, sizeof(*entry), gf_common_mt_dnscache_addr_entry);
+    return entry;
+}
+
+/**
+ * gf_dnscache_addr_entry_deinit -- Free memory used by a dnscache addr entry
+ *
+ * @entry: Pointer to deallocate
+ */
+
+void
+gf_dnscache_addr_entry_deinit(struct dnscache_addr_entry *entry)
+{
+    freeaddrinfo(entry->addr);
+    GF_FREE(entry);
+}
+
+/* Peform a dns lookup for identifier(IP address or hostname) */
+
+struct addrinfo *
+gf_dns_lookup_address_cached(char *identifier, struct dnscache *dnscache)
+{
+    struct addrinfo *addr = NULL;
+    dict_t *cache = NULL;
+    data_t *entrydata = NULL;
+    struct dnscache_addr_entry *dnsentry = NULL;
+    gf_boolean_t from_cache = _gf_false;
+    struct addrinfo hints;
+    int gai_err = 0;
+
+    if (!dnscache)
+        goto out;
+
+    cache = dnscache->cache_dict;
+
+    /* Quick cache lookup to see if we already hold it */
+    entrydata = dict_get(cache, identifier);
+    if (entrydata) {
+        dnsentry = (struct dnscache_addr_entry *)entrydata->data;
+        /* First check the TTL & timestamp */
+        if (gf_time() - dnsentry->timestamp > dnscache->ttl) {
+            gf_dnscache_addr_entry_deinit(dnsentry);
+            entrydata->data = NULL; /* Mark this as 'null' so
+                                     * dict_del () doesn't try free
+                                     * this after we've already
+                                     * freed it.
+                                     */
+
+            dict_del(cache, identifier); /* Remove this entry */
+        } else {
+            /* Cache entry is valid, get the addr and return */
+            addr = dnsentry->addr;
+            from_cache = _gf_true; /* Mark this as from cache */
+            goto out;
+        }
+    }
+
+    /* Get the addr */
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_UNSPEC;
+
+    gai_err = getaddrinfo(identifier, NULL, &hints, &addr);
+    if (gai_err != 0) {
+        gf_smsg(identifier, GF_LOG_WARNING, 0, LG_MSG_GETADDRINFO_FAILED,
+                "error=%s", gai_strerror(gai_err), NULL);
+        goto out;
+    }
+    from_cache = _gf_false;
+out:
+    /* Insert into the cache */
+    if (addr && !from_cache && identifier) {
+        struct dnscache_addr_entry *entry = gf_dnscache_addr_entry_init();
+
+        if (entry) {
```

Comment:
It's fine to not return an error if the `entry` creation fails because we already have the needed information. But in this case, who will release the memory allocated by for `addr` ?

Reply:
done